File: packages/ocel/src/blob/bucket.ts
import { ResourceType } from "../gen/proto/app/resources/v1/resources_pb.js";
import type { BucketProperties } from "../gen/proto/common/bindings/v1/bindings_pb.js";
import { declarationSite } from "../utils/callsite.js";
import { defer } from "../utils/defer.js";
import { getConfig } from "../utils/get-config.js";
import { unprovisioned, unprovisionedPhase } from "../utils/phase.js";
import { reference } from "../utils/reference.js";
import { rpc } from "../utils/rpc.js";
import type { AnyUploader } from "./types.js";

export interface BucketOptions<TUploaders extends Record<string, AnyUploader>> {
  allowedOrigins?: string[];
  uploaders: TUploaders;
}

/** The options of {@link bucket.ref}: the declaration owns the bucket's own configuration. */
export type BucketRefOptions<TUploaders extends Record<string, AnyUploader>> = Pick<
  BucketOptions<TUploaders>,
  "uploaders"
>;

export type ResolvedBucketConfig = Pick<BucketProperties, "bucket">;

export class Bucket<TUploaders extends Record<string, AnyUploader> = Record<string, AnyUploader>> {
  constructor(
    public name: string,
    public uploaders: TUploaders,
  ) {}

  __config(): ResolvedBucketConfig {
    if (unprovisionedPhase()) {
      throw unprovisioned(`bucket("${this.name}")`, "__config");
    }
    return getConfig(this.name, "bucket");
  }
}

/**
 * Declares a bucket named `name` and returns the handle its uploaders are served through.
 * Call it from a file under the project's discovery folder: during discovery the call is
 * the declaration, and at runtime it reads the binding the deploy delivered for that name.
 */
export function bucket<TUploaders extends Record<string, AnyUploader>>(
  name: string,
  options: BucketOptions<TUploaders>,
): Bucket<TUploaders> {
  if (unprovisionedPhase()) {
    defer(
      rpc.resource.declare({
        resource: { name, type: ResourceType.BUCKET },
        config: { case: "bucket", value: { allowedOrigins: options.allowedOrigins ?? [] } },
        source: declarationSite(),
      }),
    );
  }
  return new Bucket(name, options.uploaders);
}

/**
 * References the bucket named `name`, declared once elsewhere in the project in any
 * language, and returns the same handle {@link bucket} does. It never declares. Call it
 * from a file under the project's discovery folder and import that file from the app:
 * during discovery the call records that the file uses the bucket, so the deploy grants it
 * to every app that imports the file.
 */
bucket.ref = <TUploaders extends Record<string, AnyUploader>>(
  name: string,
  options: BucketRefOptions<TUploaders>,
): Bucket<TUploaders> => {
  reference(ResourceType.BUCKET, name);
  return new Bucket(name, options.uploaders);
};

File: python/ocel/src/ocel/bucket.py
import inspect
import os
from collections.abc import Sequence

from protobuf import Oneof

from ocel._binding import bucket_binding, unprovisioned
from ocel._declare import declare, discovering
from ocel.gen.app.bucket.v1.bucket_connect import BucketServiceClientSync
from ocel.gen.app.bucket.v1.bucket_pb import HeadRequest
from ocel.gen.app.resources.v1.resources_pb import (
    BucketConfig,
    DeclareRequest,
    ResourceIdentifier,
    ResourceType,
)

_KIND = "bucket"
_RUNTIME_ADDRESS_ENV = "OCEL_RUNTIME_ADDRESS"
_SESSION_TOKEN_ENV = "OCEL_SESSION_TOKEN"


class _Reached:
    def __init__(self, client: BucketServiceClientSync, bucket: str, public_base_url: str):
        self.client = client
        self.bucket = bucket
        self.public_base_url = public_base_url


class Bucket:
    """A bucket an app declares and reads and writes its objects through."""

    #: The name the bucket was declared under, and the name its binding is delivered as.
    name: str

    def __init__(self, name: str):
        """Take the handle for the bucket named ``name``. Prefer :func:`bucket`, which
        declares the bucket as well as handing back its handle."""
        self.name = name
        self._reached: _Reached | None = None

    def head(self, key: str):
        """What the bucket knows about the object under ``key``, or ``None`` when it holds
        none."""
        reached = self._runtime("head")
        reached.client.head(HeadRequest(bucket=reached.bucket, key=key))

    def _runtime(self, access: str) -> _Reached:
        if discovering():
            raise unprovisioned(f'bucket("{self.name}")', access)
        if self._reached is None:
            properties = bucket_binding(self.name)
            address = os.environ.get(_RUNTIME_ADDRESS_ENV)
            if not address:
                raise RuntimeError(
                    f"{_RUNTIME_ADDRESS_ENV} is not defined, so no resource the ocel runtime "
                    f"serves can be reached. Run `ocel dev` to serve it locally, or "
                    f"`ocel deploy` to have the deployed runtime's address delivered."
                )
            token = os.environ.get(_SESSION_TOKEN_ENV)
            if not token:
                raise RuntimeError(
                    f"{_SESSION_TOKEN_ENV} is not defined, so the ocel runtime at {address} "
                    f"would refuse every call. It is delivered beside {_RUNTIME_ADDRESS_ENV} by "
                    f"`ocel dev` and by the deployed runtime, never set by hand."
                )
            self._reached = _Reached(
                BucketServiceClientSync(address.rstrip("/"), send_compression=None),
                properties.bucket,
                properties.public_base_url,
            )
        return self._reached


def bucket(name: str, *, public: bool = False, allowed_origins: Sequence[str] = ()) -> Bucket:
    """Declare a bucket named ``name`` and return the handle an app reads and writes its
    objects through. Call it from a file under the project's discovery folder: during
    discovery the call is the declaration, and at runtime it reads the binding the deploy
    delivered for that name. A ``public`` bucket serves every object it holds anonymously
    over HTTP, and ``allowed_origins`` names the browser origins allowed to upload straight
    to the store."""
    if not discovering():
        return Bucket(name)
    caller = inspect.stack(0)[1]
    declare(
        DeclareRequest(
            resource=ResourceIdentifier(type=ResourceType.BUCKET, name=name),
            config=Oneof(_KIND, BucketConfig(public=public, allowed_origins=list(allowed_origins))),
            source=f"{caller.filename}:{caller.lineno}",
        )
    )
    return Bucket(name)
